.github/workflows/build_e2e_test_artifacts.yml

```diff
@@ -39,6 +39,14 @@ on:
           if set to 1.
         default: 1
         type: number
+      shard-count:
+        description: |
+          A device-name to integer mapping as a comma separated list.
+          Allows to assign a distinct shard count for each target device.
+          The reserved keyword `default` assigns a shard count to all target devices
+          that are not explicitly listed.
+        default: default=1
```

Comment:
Why is the default here different from the default in the other file?

Reply:
Ok this makes sense. So this controls how many shards the execution config gets generated with. It's invoked from the main workflow where it doesn't specify the shard count and so gets this default, but when it's invoked from the large workflow that asks for sharding of the desktop benchmarks (is that actually all the benchmarks for "large"? Should it just set `default=2`?). One thing that's a little odd to me is that this is a parameter to the *building* of artifacts. To me that suggests that the building would be sharded, but it's not: it just generates some configs that do sharding. I wonder if resharding should instead be done as a separate step.

One alternative to consider is to match how Bazel does test sharding: https://bazel.build/reference/test-encyclopedia#test-sharding. We'd give every shard the full config and also tell it how many shards there are and what index it is. Then it's responsible for locally computing which it should run. Your approach has the advantage that we can do global optimization of the sharding rather than being limited only to simple things that can be computed locally, like round robin. How likely are we to go beyond round robin though? And if we do, could we just reorder the modules in the config? If we were to actually go for much more intelligent sharding I think I would lean towards dynamic scheduling with workers and tasks and in that case a static config wouldn't work anyway.

Another thing to consider is that I might want to execute the same modules but sharded in a different way. That makes it a bit awkward that sharding is part of the config.

Anyways, just some things to consider. Not blocking if you prefer this method.

Reply:
From another point of view, we actually don't need to generate `execution_benchmark_config.json` in `build_e2e_test_artifacts` and it wasn't. I moved it from `benchmark_execution.yml` into `build_e2e_test_artifacts.yml` simply because I was trying to make sure all benchmark related artifacts including configs can be download from `gcs://..../e2e-test-artifacts` for users. The problem to generate and upload configs from `benchmark_execution.yml` is that it can be under a different GCS URL during the rerun, because we upload artifacts to different GCS root dir in different run attempts.

I think fundamentally the problem is that `execution_benchmark_config.json` is playing too many roles currently. It provides the benchmark metadata (flags, tags) for searchability and reproducibility but benchmark tools also read it to know which benchmarks to run. So the benchmarks in the config are grouped by target devices (and shards), which is unnecessary if we have another channel to tell tools which benchmarks to run from the config...

Reply:
> 



> Ok this makes sense. So this controls how many shards the execution config gets generated with. It's invoked from the main workflow where it doesn't specify the shard count and so gets this default, but when it's invoked from the large workflow that asks for sharding of the desktop benchmarks (is that actually all the benchmarks for "large"? Should it just set `default=2`?). 

I considered making "default=2" the default value, but I figured at some point someone might need to change that value (either for a one-off run or in general), and then `a2-highgpu-1g=2,c2-standard-16=2` is kind of self explanatory. They will know what to do without looking into the code. I also expect that we might add new machine types at some point (a2-highgpu-8g or even a3-*) - starting out with a very limited set of benchmarks where we don't won't sharding yet. Of course that's just a guess.

>One thing that's a little odd to me is that this is a parameter to the _building_ of artifacts. To me that suggests that the building would be sharded, but it's not: it just generates some configs that do sharding. I wonder if resharding should instead be done as a separate step.

Yes, I followed this approach first and that quickly hit a couple of obstacles that would have made quite a bit more involved. For example we have a step for downloading benchmark assets on the benchmark-executing machine. This step is currently just a shell command (`jq` read the benchmark execution config and pipes a list of GCS urls into `gcloud storage`.). If we wanted to have sharding out of the execution config, then this step would need to know about how we shard whcih probably means moving this code into Python and doing a bit of refactoring in Python as well.

So I ended up following Che-Yu's advice and put the sharding into the execution config.

> 
> One alternative to consider is to match how Bazel does test sharding: https://bazel.build/reference/test-encyclopedia#test-sharding. We'd give every shard the full config and also tell it how many shards there are and what index it is. Then it's responsible for locally computing which it should run. Your approach has the advantage that we can do global optimization of the sharding rather than being limited only to simple things that can be computed locally, like round robin. How likely are we to go beyond round robin though? And if we do, could we just reorder the modules in the config? If we were to actually go for much more intelligent sharding I think I would lean towards dynamic scheduling with workers and tasks and in that case a static config wouldn't work anyway.

For comparative benchmarks we want to run the same benchmarks with different compilers on the same machine, so this will probably go beyond round robin. Also I've seen that round robin can lead to quite an unbalanced use of the shards, so I considered introducing some simple scalar which gauges the expected compute needed to run this benchmark (I assume batch size is already a good enough indicator) and then shard with this heuristic in mind.

Neverthess technically this could all be done in a local manner on each machine. As long as the sharding strategy is stable we are not bound to doing this globally.

> 
> Another thing to consider is that I might want to execute the same modules but sharded in a different way. That makes it a bit awkward that sharding is part of the config.

Are there plans to support that on the CI? I think currently the modules are always regenerated and so is the config. And the latter can have a different sharding config.

In general I see 2 ways to approach this class of problems:

1. We could always upload 2 config files - one with sharding which the CI uses and one without sharding which anybody could use for anything
2. We can provide a simple python helper which reads an existing config file and allows to re-shard it in a different way (different strategy and/or different shard count). Of course this would share the code with the existing export_benchmark_config script.

> 
> Anyways, just some things to consider. Not blocking if you prefer this method.

Reply:
...and I just realized that Che-Yu basically answered the same questions as well. Sorry for the repetition.